_source/_posts/2019-11-19-secure-kafka-java.md

```diff
@@ -0,0 +1,458 @@
+
+That's it! Now your Java project structure is created, and you can start developing your app.
+
+## Push Messages to a Kafka Topic in Your Java App
+
+The first step to create a producer that can push messages is to configure the producers inside your Java application. Let's create a configuration class to do just that.
+
+Create a `src/main/java/com/okta/javakafka/configuration` folder, and a `ProducerConfiguration` class in it:
+
+```java
+import org.apache.kafka.clients.producer.ProducerConfig;
+import org.apache.kafka.common.serialization.StringSerializer;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.kafka.core.DefaultKafkaProducerFactory;
+import org.springframework.kafka.core.KafkaTemplate;
+import org.springframework.kafka.core.ProducerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+
+@Configuration
+public class ProducerConfiguration {
+
+    private static final String KAFKA_BROKER = "localhost:9092";
+
+    @Bean
+    public ProducerFactory<String, String> producerFactory() {
+        return new DefaultKafkaProducerFactory<>(producerConfigurations());
+    }
+
+    @Bean
+    public Map<String, Object> producerConfigurations() {
+        Map<String, Object> configurations = new HashMap<>();
+
+        configurations.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, KAFKA_BROKER);
+        configurations.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
+        configurations.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
+
+        return configurations;
+    }
+
+    @Bean
+    public KafkaTemplate<String, String> kafkaTemplate() {
+        return new KafkaTemplate<>(producerFactory());
+    }
+
+}
+```
+
+This class creates a `ProducerFactory` which knows how to create producers based on the configurations you provided. You've also specified to connect to your local Kafka broker and to serialize both the key and the values with `String`.
+
+You also declared a `KafkaTemplate` bean to perform high-level operations on your producer. In other words, the template is able to do operations such as sending a message to a topic and efficiently hides under-the-hood details from you.
+
+The next step is to create the endpoint to send the message to the producer. Inside the `src/main/java/com/okta/javakafka/controller` package, create the following class:
+
+```java
+import org.springframework.kafka.core.KafkaTemplate;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+
+import java.util.List;
+
+@RestController
+public class KafkaController {
+
+    private KafkaTemplate<String, String> template;
+
+    public KafkaController(KafkaTemplate<String, String> template) {
+        this.template = template;
+    }
+
+    @GetMapping("/kafka/produce")
```

Comment:
Since you're sending data to be processed, this really ought to be a POST. BUT, you wouldn't be able to exercise it from the browser as described in the blog post.

Maybe:
(1) change it to a POST and alter the blog post to use curl instead of the browser OR
(2) leave it as a GET, but put some sort of disclaimer in the blog post that "in real life" this would be a POST, but for demo purposes it's easier to leave it as a GET so you can exercise it in the browser.

Reply:
I'll add a note.